pallets/asset-manager/src/migrations.rs

```diff
@@ -0,0 +1,416 @@
+// Copyright 2019-2021 PureStake Inc.
+// This file is part of Moonbeam.
+
+// Moonbeam is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Moonbeam is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Moonbeam.  If not, see <http://www.gnu.org/licenses/>.
+
+use crate::{AssetIdType, AssetTypeId, AssetTypeUnitsPerSecond, Config};
+use frame_support::{
+	pallet_prelude::PhantomData,
+	storage::migration::{remove_storage_prefix, storage_key_iter},
+	traits::{Get, OnRuntimeUpgrade},
+	weights::Weight,
+	Blake2_128Concat,
+};
+use sp_std::convert::TryInto;
+//TODO sometimes this is unused, sometimes its necessary
+use sp_std::vec::Vec;
+use xcm::latest::prelude::*;
+
+pub struct UnitsWithAssetType<T>(PhantomData<T>);
+impl<T: Config> OnRuntimeUpgrade for UnitsWithAssetType<T> {
+	fn on_runtime_upgrade() -> Weight {
+		log::info!(target: "UnitsWithAssetType", "actually running it");
+		let pallet_prefix: &[u8] = b"AssetManager";
+		let storage_item_prefix: &[u8] = b"AssetIdUnitsPerSecond";
+
+		// Read all the data into memory.
+		// https://crates.parity.io/frame_support/storage/migration/fn.storage_key_iter.html
+		let stored_data: Vec<_> = storage_key_iter::<T::AssetId, u128, Blake2_128Concat>(
+			pallet_prefix,
+			storage_item_prefix,
+		)
+		.collect();
+
+		let migrated_count: Weight = stored_data
+			.len()
+			.try_into()
+			.expect("There are between 0 and 2**64 mappings stored.");
+
+		log::info!(target: "UnitsWithAssetType", "Migrating {:?} elements", migrated_count);
+
+		// Now remove the old storage
+		// https://crates.parity.io/frame_support/storage/migration/fn.remove_storage_prefix.html
+		remove_storage_prefix(pallet_prefix, storage_item_prefix, &[]);
```

Comment:
You might be able to use `drain()` instead of `collect()` above to accomplish the same. Not a big deal at all though...